pandas/io/parquet.py

```diff
@@ -180,6 +180,9 @@ def write(
             mode="wb",
             is_dir=partition_cols is not None,
         )
+        if isinstance(path_or_handle, WriteBuffer):
```

Comment:
`pandas._typing.Writebuffer` and  `io.BufferWriter` are two different things. Classes that implement the protocol `pandas._typing.Writebuffer` might not have `raw.name`.

Reply:
Oh got it thanks, let me change the io.BufferWriter class.

Reply:
I'm not sure whether your approach is an "ideal" solution. Ideally, we want to detect that we will encounter an unsupported operation and then raise (before doing any IO).

Reply:
yes, ideally we would like to capture the error (in this case an unsupported dtype for to_parquet function using pyarrow engine) before io. PyArrow raises an exception when it encounters an unsupported dtype, but before it raises the exception itself it performs a partial flush to disk.

We were unable to replicate the issue directly on PyArrow, and it turns out that PyArrow handles the removal of the partial file only when the file path is a string or path like object (https://github.com/apache/arrow/blob/657d21ad6e56cc6c2c9b1aeeb5cea983210296b6/python/pyarrow/parquet.py#L2093)

Adding an explicit check on dtypes instead of letting PyArrow handle it would mean that any new improvements in PyArrow wouldn't be captured in Pandas. (https://github.com/pandas-dev/pandas/pull/44847#issuecomment-992440761)

Earlier discussions on this issue: https://github.com/pandas-dev/pandas/pull/44847